### PR TITLE
Impl-multus

### LINKED ISSUE #103: Allow to deploy multus in addition to another CNI plugin
**Describe the solution you'd like:**

I'd like to allow deployment of multus cni alongside with another cni provider, it is neither supported [by current API](https://github.com/rancher-sandbox/cluster-api-provider-rke2/blob/v0.1.0-alpha.1/controlplane/api/v1alpha1/rke2controlplane_types.go#L95-L98) nor implemented.

Current API description is misleading as it claims that
```  
     CNI describes the CNI Plugins to deploy, one of none, calico, canal,
     cilium; optionally with multus as the first value to enable the multus
     meta-plugin (default: canal).
```  
Whereas the schema won't accept a value like `cni: multus,calico`

### PR DESCRIPTION
**Pull request by belgaied2**
Signed-off-by: Mohamed Belgaied Hassine <belgaied2@hotmail.com>

**What this PR does / why we need it**:
This PR implements a way to activate multus as a primary CNI, making the plugin defined in the CNI field secondary CNI.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #103

**Special notes for your reviewer**:

**Checklist**:

- [X] squashed commits into logical changes
- [ ] includes documentation
- [ ] adds unit tests
- [ ] adds or updates e2e tests

### CODE REVIEW

File: samples/docker/enable-multus/rke2controlplane-test.yaml [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+metadata:
+  name: rke2cp-test
+  namespace: webhook-test
+spec: 
+  replicas: 3
+  agentConfig:
+    version: v1.24.11+rke2r1
+  serverConfig:
+    cniMultusEnable: true
+      #cni: calico
```
Comment (furkatgofurov7) [inline samples/docker/enable-multus/rke2controlplane-test.yaml:12]: this should be uncommented, right?

File: controlplane/api/v1alpha1/rke2controlplane_types.go [hunk 2]
```diff
@@ -102,6 +102,11 @@ type RKE2ServerConfig struct {
 	//+optional
 	CNI CNI `json:"cni,omitempty"`
 
+	// CNIMultusEnable enables multus as the first CNI plugin (default: false).
+	// This option will automatically overtake the CNI option as secondary CNI plugin.
```
Comment (furkatgofurov7) [inline controlplane/api/v1alpha1/rke2controlplane_types.go:106]: This statement is confusing at least to me. If we enable Multus CNI as a primary CNI via this field, why it is considered as secondary CNI plugin?

**Review by furkatgofurov7 — COMMENTED** [review]
I feel like we are tackling two different issues like choosing the CNI & disableComponents fields fixes (which actually points to the closed issue) in the same PR. It would be great if possible to separate out the second commit onto it's own PR, so that it is easier reviewing.

File: controlplane/api/v1alpha1/rke2controlplane_types.go [hunk 3]
```diff
@@ -293,15 +298,14 @@ const (
 // DisableComponents describes components of RKE2 (Kubernetes components and plugin components) that should be disabled.
 type DisableComponents struct {
 	// KubernetesComponents is a list of Kubernetes components to disable.
-	// +kubebuilder:validation:Enum=scheduler;kubeProxy;cloudController
 	KubernetesComponents []DisabledKubernetesComponent `json:"kubernetesComponents,omitempty"`
 
 	// PluginComponents is a list of PluginComponents to disable.
-	// +kubebuilder:validation:Enum=rke2-coredns;rke2-ingress-nginx;rke2-metrics-server
```
[the remark this answers is not in the document]
Comment (belgaied2) [reply controlplane/api/v1alpha1/rke2controlplane_types.go:300]: This is actually part of another PR: #130. This PR was done on top of the other commits.

File: controlplane/api/v1alpha1/rke2controlplane_types.go (hunk 2 above)
Comment (alexander-demicev) [reply controlplane/api/v1alpha1/rke2controlplane_types.go:106]: Maybe we can create CNIPlugins field? In case we'll want to extend the list later

File: controlplane/internal/controllers/rke2controlplane_controller.go [hunk 4]
```diff
@@ -649,8 +648,23 @@ func (r *RKE2ControlPlaneReconciler) reconcileKubeconfig(
 func (r *RKE2ControlPlaneReconciler) reconcileControlPlaneConditions(ctx context.Context, controlPlane *rke2.ControlPlane) (ctrl.Result, error) {
 	logger := log.FromContext(ctx)
 
-	if controlPlane.Machines.Len() == 0 {
+	readyCPMachines := controlPlane.Machines.Filter(collections.IsReady())
+
+	if readyCPMachines.Len() == 0 {
```
Comment (alexander-demicev) [inline controlplane/internal/controllers/rke2controlplane_controller.go:653]: why are we erasing all other conditions here?

File: controlplane/internal/controllers/rke2controlplane_controller.go [hunk 5]
```diff
@@ -346,7 +346,8 @@ func (r *RKE2ControlPlaneReconciler) updateStatus(ctx context.Context, rcp *cont
 		return fmt.Errorf("no Control Plane Machines are ready for RKE2ControlPlane %s/%s", rcp.Namespace, rcp.Name)
 	}
 
-	availableCPMachines := readyMachines.Filter(collections.Not(collections.HasUnhealthyCondition))
+	availableCPMachines := readyMachines
+	//.Filter(collections.Not(collections.HasUnhealthyCondition))
```
[the remark this answers is not in the document]
Comment (belgaied2) [reply controlplane/internal/controllers/rke2controlplane_controller.go:350]: Well, this is part of an improvement that I mentioned previously to @alexander-demicev and @richardcase. If we only select Healthy machines, we constraint ourself too much because of the way CAPI defines a healthy machine.

For instance: on AWS, if you deploy the external Cloud Provider, the machine will only be healthy if the cloud provider populates the `providerID` field in the workload cluster's node. If, for example, your Cloud Provider does not deploy on your first node because your first node has a `ControlPlane` taint, the machine will never be healthy, and will therefore never be added to the list of available IPs for registration of other nodes.

RKE2 does not need this kind of behaviour. You can make additional nodes register with the init nodes when these node is `Ready`, not necessarily when it is healthy and with a valid `ProviderID` field.

File: samples/docker/enable-multus/rke2controlplane-test.yaml (hunk 1 above)
Comment (belgaied2) [reply samples/docker/enable-multus/rke2controlplane-test.yaml:12]: Well, both are fine. Maybe for our users I should uncomment, but an empty `cni` field with `cniMultusEnable` to `true` was an edge case that I tested.

File: controlplane/api/v1alpha1/rke2controlplane_types.go (hunk 2 above)
Comment (belgaied2) [reply controlplane/api/v1alpha1/rke2controlplane_types.go:106]: Well, I think the godoc is unclear.  I might something like : 

```go
// This option will automatically make Multus primary CNI, and whatever mentioned in the cni field as secondary CNI plugin.
```

Would that work better or do you have a better formulation ?

File: controlplane/api/v1alpha1/rke2controlplane_types.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (furkatgofurov7) [reply controlplane/api/v1alpha1/rke2controlplane_types.go:300]: Yes that is what I guessed as well, and why not to separate out it from this PR?

File: controlplane/internal/controllers/rke2controlplane_controller.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (furkatgofurov7) [reply controlplane/internal/controllers/rke2controlplane_controller.go:350]: well the cp taint is users responsibility to put or leave out right

File: samples/docker/enable-multus/rke2controlplane-test.yaml (hunk 1 above)
Comment (furkatgofurov7) [reply samples/docker/enable-multus/rke2controlplane-test.yaml:12]: I was just making sure we either remove or uncomment it not to confuse the users

File: controlplane/api/v1alpha1/rke2controlplane_types.go (hunk 2 above)
Comment (furkatgofurov7) [reply controlplane/api/v1alpha1/rke2controlplane_types.go:106]: maybe:

`// This option will automatically make Multus a primary CNI, and the value, if specified in the CNI field, as a secondary CNI plugin.
`

File: controlplane/api/v1alpha1/rke2controlplane_types.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (belgaied2) [reply controlplane/api/v1alpha1/rke2controlplane_types.go]: Since I did both PRs one after the other, I didn't manage well my branching locally... Now, I removed the commit above and force-pushed again.

File: controlplane/api/v1alpha1/rke2controlplane_webhook.go [hunk 6]
```diff
@@ -64,3 +74,25 @@ func (r *RKE2ControlPlane) ValidateDelete() error {
...
+func ValidateRKE2ControlPlaneSpec(name string, spec *RKE2ControlPlaneSpec) error {
+	allErrs := spec.validate()
+	if len(allErrs) == 0 {
+		return nil
+	}
+
+	return apierrors.NewInvalid(GroupVersion.WithKind("RKE2ControlPlane").GroupKind(), name, allErrs)
+}
+
+func (s *RKE2ControlPlaneSpec) validate() field.ErrorList {
```
[the remark this answers is not in the document]
Comment (belgaied2) [reply controlplane/api/v1alpha1/rke2controlplane_webhook.go:89]: Added! :)

File: controlplane/api/v1alpha1/rke2controlplane_types.go (hunk 2 above)
Comment (belgaied2) [reply controlplane/api/v1alpha1/rke2controlplane_types.go:106]: Sounds good, I just did that.

File: samples/docker/enable-multus/rke2controlplane-test.yaml (hunk 1 above)
Comment (belgaied2) [reply samples/docker/enable-multus/rke2controlplane-test.yaml]: Done. :)

File: controlplane/api/v1alpha1/rke2controlplane_types.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (furkatgofurov7) [reply controlplane/api/v1alpha1/rke2controlplane_types.go]: Thanks, it is much cleaner now!

File: controlplane/internal/controllers/rke2controlplane_controller.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (furkatgofurov7) [reply controlplane/internal/controllers/rke2controlplane_controller.go:350]: Thanks for putting together the detailed explanation, it generally makes sense. Please remove the leftover commented code

**Review by furkatgofurov7 — APPROVED** [review]
Thank you @belgaied2, this looks good to me.

**Review by furkatgofurov7 — APPROVED** [review]
